src/librustc_middle/ty/sty.rs

```diff
@@ -348,9 +354,20 @@ impl<'tcx> ClosureSubsts<'tcx> {
         self.substs.len() >= 3 && matches!(self.split().tupled_upvars_ty.expect_ty().kind, Tuple(_))
     }
 
+    /// Returns the substitutions of the closure's parent.
+    pub fn parent_substs(self) -> &'tcx [GenericArg<'tcx>] {
```

Comment:
All of these changes as well, can be kept in the PR that uses them.

Reply:
Fixed